// @tanstack/start-plugin-core/dist/esm/import-protection-plugin/postCompileUsage.js
import * as t from "@babel/types";
import { parseAst } from "@tanstack/router-utils";
function collectPatternBindings(node, out) {
  if (!node) return;
  if (t.isIdentifier(node)) {
    out.add(node.name);
    return;
  }
  if (t.isRestElement(node)) {
    collectPatternBindings(node.argument, out);
    return;
  }
  if (t.isAssignmentPattern(node)) {
    collectPatternBindings(node.left, out);
    return;
  }
  if (t.isObjectPattern(node)) {
    for (const prop of node.properties) {
      if (t.isRestElement(prop)) {
        collectPatternBindings(prop.argument, out);
      } else if (t.isObjectProperty(prop)) {
        collectPatternBindings(prop.value, out);
      }
    }
    return;
  }
  if (t.isArrayPattern(node)) {
    for (const el of node.elements) {
      collectPatternBindings(el, out);
    }
    return;
  }
}
function isBindingPosition(node, parent) {
  if (!parent) return false;
  if (t.isFunctionDeclaration(parent) && parent.id === node) return true;
  if (t.isFunctionExpression(parent) && parent.id === node) return true;
  if (t.isClassDeclaration(parent) && parent.id === node) return true;
  if (t.isClassExpression(parent) && parent.id === node) return true;
  if (t.isVariableDeclarator(parent) && parent.id === node) return true;
  if (t.isImportSpecifier(parent) && parent.local === node) return true;
  if (t.isImportDefaultSpecifier(parent) && parent.local === node) return true;
  if (t.isImportNamespaceSpecifier(parent) && parent.local === node) return true;
  if (t.isObjectProperty(parent) && parent.key === node && !parent.computed && // In `{ foo }`, the identifier is also a value reference and must count as
  // usage. Babel represents this as `shorthand: true`.
  !parent.shorthand)
    return true;
  if (t.isObjectMethod(parent) && parent.key === node && !parent.computed)
    return true;
  if (t.isExportSpecifier(parent) && parent.exported === node) return true;
  return false;
}
function isPreferredUsage(node, parent) {
  if (!parent) return false;
  if (t.isCallExpression(parent) && parent.callee === node) return true;
  if (t.isNewExpression(parent) && parent.callee === node) return true;
  if (t.isMemberExpression(parent) && parent.object === node) return true;
  return false;
}
function isScopeNode(node) {
  return t.isProgram(node) || t.isFunctionDeclaration(node) || t.isFunctionExpression(node) || t.isArrowFunctionExpression(node) || t.isBlockStatement(node) || t.isCatchClause(node);
}
function isFunctionScopeNode(node) {
  return t.isProgram(node) || t.isFunctionDeclaration(node) || t.isFunctionExpression(node) || t.isArrowFunctionExpression(node);
}
function collectScopeBindings(node, out) {
  if (t.isFunctionDeclaration(node) || t.isFunctionExpression(node) || t.isArrowFunctionExpression(node)) {
    for (const p of node.params) {
      collectPatternBindings(p, out);
    }
    return;
  }
  if (t.isCatchClause(node)) {
    collectPatternBindings(node.param, out);
    return;
  }
}
function findPostCompileUsagePos(code, source) {
  const ast = parseAst({ code });
  const imported = /* @__PURE__ */ new Set();
  for (const node of ast.program.body) {
    if (t.isImportDeclaration(node) && node.source.value === source) {
      if (node.importKind === "type") continue;
      for (const s of node.specifiers) {
        if (t.isImportSpecifier(s) && s.importKind === "type") continue;
        imported.add(s.local.name);
      }
    }
  }
  if (imported.size === 0) return void 0;
  let preferred;
  let anyUsage;
  const scopes = [{ bindings: /* @__PURE__ */ new Set(), isFnScope: true }];
  function isShadowed(name) {
    for (let i = scopes.length - 1; i >= 1; i--) {
      if (scopes[i].bindings.has(name)) return true;
    }
    return false;
  }
  function record(node, kind) {
    const loc = node.loc?.start;
    if (!loc) return;
    const pos = { line: loc.line, column0: loc.column };
    if (kind === "preferred") {
      preferred ||= pos;
    } else {
      anyUsage ||= pos;
    }
  }
  function pushScope(node) {
    const bindings = /* @__PURE__ */ new Set();
    collectScopeBindings(node, bindings);
    scopes.push({ bindings, isFnScope: isFunctionScopeNode(node) });
  }
  function popScope() {
    scopes.pop();
  }
  function nearestFnScope() {
    for (let i = scopes.length - 1; i >= 0; i--) {
      if (scopes[i].isFnScope) return scopes[i];
    }
    return scopes[0];
  }
  function walk(node, parent) {
    if (!node) return;
    if (preferred && anyUsage) return;
    if (Array.isArray(node)) {
      for (const n of node) walk(n, parent);
      return;
    }
    const astNode = node;
    if (t.isImportDeclaration(astNode)) return;
    const enterScope = isScopeNode(astNode);
    if (enterScope) {
      pushScope(astNode);
    }
    if (t.isFunctionDeclaration(astNode) && astNode.id) {
      scopes[scopes.length - 2]?.bindings.add(astNode.id.name);
    }
    if (t.isClassDeclaration(astNode) && astNode.id) {
      scopes[scopes.length - 2]?.bindings.add(astNode.id.name);
    }
    if (t.isVariableDeclarator(astNode)) {
      const isVar = t.isVariableDeclaration(parent) && parent.kind === "var";
      const target = isVar ? nearestFnScope().bindings : scopes[scopes.length - 1].bindings;
      collectPatternBindings(astNode.id, target);
    }
    if (t.isIdentifier(astNode) && imported.has(astNode.name)) {
      if (!isBindingPosition(astNode, parent) && !isShadowed(astNode.name)) {
        if (isPreferredUsage(astNode, parent)) {
          record(astNode, "preferred");
        } else {
          record(astNode, "any");
        }
      }
    }
    const record_ = astNode;
    for (const key of Object.keys(record_)) {
      const value = record_[key];
      if (!value) continue;
      if (key === "loc" || key === "start" || key === "end") continue;
      if (key === "parent") continue;
      if (typeof value === "string" || typeof value === "number") continue;
      walk(value, astNode);
      if (preferred && anyUsage) break;
    }
    if (enterScope) {
      popScope();
    }
  }
  walk(ast.program, null);
  return preferred ?? anyUsage;
}
export {
  findPostCompileUsagePos
};
