// @tanstack/start-plugin-core/dist/esm/import-protection-plugin/sourceLocation.js
import { SourceMapConsumer } from "source-map";
import * as path from "pathe";
import { normalizeFilePath } from "./utils.js";
function buildLineIndex(code) {
  const offsets = [0];
  for (let i = 0; i < code.length; i++) {
    if (code.charCodeAt(i) === 10) {
      offsets.push(i + 1);
    }
  }
  return { offsets };
}
function upperBound(values, x) {
  let lo = 0;
  let hi = values.length;
  while (lo < hi) {
    const mid = lo + hi >> 1;
    if (values[mid] <= x) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}
function indexToLineColWithIndex(lineIndex, idx) {
  let line = 1;
  const offsets = lineIndex.offsets;
  const ub = upperBound(offsets, idx);
  const lineIdx = Math.max(0, ub - 1);
  line = lineIdx + 1;
  const lineStart = offsets[lineIdx] ?? 0;
  return { line, column0: Math.max(0, idx - lineStart) };
}
function suffixSegmentScore(a, b) {
  const aSeg = a.split("/").filter(Boolean);
  const bSeg = b.split("/").filter(Boolean);
  let score = 0;
  for (let i = aSeg.length - 1, j = bSeg.length - 1; i >= 0 && j >= 0; i--, j--) {
    if (aSeg[i] !== bSeg[j]) break;
    score++;
  }
  return score;
}
function normalizeSourceCandidate(source, root, sourceRoot) {
  if (!source) return "";
  if (path.isAbsolute(source)) return normalizeFilePath(source);
  const base = sourceRoot ? path.resolve(root, sourceRoot) : root;
  return normalizeFilePath(path.resolve(base, source));
}
function pickOriginalCodeFromSourcesContent(map, importerFile, root) {
  if (!map?.sourcesContent || map.sources.length === 0) {
    return void 0;
  }
  const file = normalizeFilePath(importerFile);
  const sourceRoot = map.sourceRoot;
  let bestIdx = -1;
  let bestScore = -1;
  for (let i = 0; i < map.sources.length; i++) {
    const content = map.sourcesContent[i];
    if (typeof content !== "string") continue;
    const src = map.sources[i] ?? "";
    const normalizedSrc = normalizeFilePath(src);
    if (normalizedSrc === file) {
      return content;
    }
    const resolved = normalizeSourceCandidate(src, root, sourceRoot);
    if (resolved === file) {
      return content;
    }
    const score = Math.max(
      suffixSegmentScore(normalizedSrc, file),
      suffixSegmentScore(resolved, file)
    );
    if (score > bestScore) {
      bestScore = score;
      bestIdx = i;
    }
  }
  if (bestIdx !== -1 && bestScore >= 1) {
    const best = map.sourcesContent[bestIdx];
    return typeof best === "string" ? best : void 0;
  }
  const fallback = map.sourcesContent[0];
  return typeof fallback === "string" ? fallback : void 0;
}
async function mapGeneratedToOriginal(map, generated, fallbackFile) {
  const fallback = {
    file: fallbackFile,
    line: generated.line,
    column: generated.column0 + 1
  };
  if (!map) {
    return fallback;
  }
  const consumer = await getSourceMapConsumer(map);
  if (!consumer) return fallback;
  try {
    const orig = consumer.originalPositionFor({
      line: generated.line,
      column: generated.column0
    });
    if (orig.line != null && orig.column != null) {
      return {
        file: orig.source ? normalizeFilePath(orig.source) : fallbackFile,
        line: orig.line,
        column: orig.column + 1
      };
    }
  } catch {
  }
  return fallback;
}
const consumerCache = /* @__PURE__ */ new WeakMap();
async function getSourceMapConsumer(map) {
  const cached = consumerCache.get(map);
  if (cached) return cached;
  const promise = (async () => {
    try {
      return await new SourceMapConsumer(map);
    } catch {
      return null;
    }
  })();
  consumerCache.set(map, promise);
  return promise;
}
function findFirstImportSpecifierIndex(code, source) {
  const escaped = source.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const patterns = [
    // import 'x'
    new RegExp(`\\bimport\\s+(['"])${escaped}\\1`),
    // import ... from 'x' / export ... from 'x'
    new RegExp(`\\bfrom\\s+(['"])${escaped}\\1`),
    // import('x')
    new RegExp(`\\bimport\\s*\\(\\s*(['"])${escaped}\\1\\s*\\)`)
  ];
  let best = -1;
  for (const re of patterns) {
    const m = re.exec(code);
    if (!m) continue;
    const idx = m.index + m[0].indexOf(source);
    if (idx === -1) continue;
    if (best === -1 || idx < best) best = idx;
  }
  return best;
}
async function findImportStatementLocationFromTransformed(provider, importerId, source, importLocCache) {
  const importerFile = normalizeFilePath(importerId);
  const cacheKey = `${importerFile}::${source}`;
  if (importLocCache.has(cacheKey)) {
    return importLocCache.get(cacheKey) || void 0;
  }
  try {
    const res = provider.getTransformResult(importerId);
    if (!res) {
      importLocCache.set(cacheKey, null);
      return void 0;
    }
    const { code, map } = res;
    if (typeof code !== "string") {
      importLocCache.set(cacheKey, null);
      return void 0;
    }
    const lineIndex = res.lineIndex ?? buildLineIndex(code);
    const idx = findFirstImportSpecifierIndex(code, source);
    if (idx === -1) {
      importLocCache.set(cacheKey, null);
      return void 0;
    }
    const generated = indexToLineColWithIndex(lineIndex, idx);
    const loc = await mapGeneratedToOriginal(map, generated, importerFile);
    importLocCache.set(cacheKey, loc);
    return loc;
  } catch {
    importLocCache.set(cacheKey, null);
    return void 0;
  }
}
async function findPostCompileUsageLocation(provider, importerId, source, findPostCompileUsagePos) {
  try {
    const importerFile = normalizeFilePath(importerId);
    const res = provider.getTransformResult(importerId);
    if (!res) return void 0;
    const { code, map } = res;
    if (typeof code !== "string") return void 0;
    if (!res.lineIndex) {
      res.lineIndex = buildLineIndex(code);
    }
    const pos = findPostCompileUsagePos(code, source);
    if (!pos) return void 0;
    return await mapGeneratedToOriginal(map, pos, importerFile);
  } catch {
    return void 0;
  }
}
async function addTraceImportLocations(provider, trace, importLocCache) {
  for (const step of trace) {
    if (!step.specifier) continue;
    if (step.line != null && step.column != null) continue;
    const loc = await findImportStatementLocationFromTransformed(
      provider,
      step.file,
      step.specifier,
      importLocCache
    );
    if (!loc) continue;
    step.line = loc.line;
    step.column = loc.column;
  }
}
function buildCodeSnippet(provider, moduleId, loc, contextLines = 2) {
  try {
    const importerFile = normalizeFilePath(moduleId);
    const res = provider.getTransformResult(moduleId);
    if (!res) return void 0;
    const { code: transformedCode, originalCode } = res;
    if (typeof transformedCode !== "string") return void 0;
    const sourceCode = originalCode ?? transformedCode;
    const allLines = sourceCode.split(/\r?\n/);
    const targetLine = loc.line;
    const targetCol = loc.column;
    if (targetLine < 1 || targetLine > allLines.length) return void 0;
    const startLine = Math.max(1, targetLine - contextLines);
    const endLine = Math.min(allLines.length, targetLine + contextLines);
    const gutterWidth = String(endLine).length;
    const sourceFile = loc.file ?? importerFile;
    const snippetLines = [];
    for (let i = startLine; i <= endLine; i++) {
      const lineContent = allLines[i - 1] ?? "";
      const lineNum = String(i).padStart(gutterWidth, " ");
      const marker = i === targetLine ? ">" : " ";
      snippetLines.push(`  ${marker} ${lineNum} | ${lineContent}`);
      if (i === targetLine && targetCol > 0) {
        const padding = " ".repeat(targetCol - 1);
        snippetLines.push(`    ${" ".repeat(gutterWidth)} | ${padding}^`);
      }
    }
    return {
      lines: snippetLines,
      highlightLine: targetLine,
      location: `${sourceFile}:${targetLine}:${targetCol}`
    };
  } catch {
    return void 0;
  }
}
export {
  addTraceImportLocations,
  buildCodeSnippet,
  buildLineIndex,
  findFirstImportSpecifierIndex,
  findImportStatementLocationFromTransformed,
  findPostCompileUsageLocation,
  mapGeneratedToOriginal,
  pickOriginalCodeFromSourcesContent
};
